### ISSUE TITLE
UnhandledPromiseRejectionWarning in Simulation

### ISSUE THREAD

**Issue by MarioHenze** [carlos8f/zenbot#1972]
### System information
- **Have I written custom code (as opposed to using zenbot vanilla)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu Server 18.04 LTS
- **Zenbot version** (commit ref, or version): 5ff40adad9cce76ab1b4ed7f88df5823bbcb32d0
- **Zenbot branch**: unstable
- **NodeJS version**: V10.18.0
- **Python version (when using a python script)**: ---
- **Exact command to reproduce (include everything)**: zenbot sim gdax.ZRX-EUR
- **Did I make any changes to conf-sample.js?**: No

### Describe the problem

Simulation sometimes leads to unhandled exceptions as described in the log below. Unfortunately this behavior was not easily reproducible. At times the bash output just stopped. I suspect the commit 42cace9323bcf2c638f90c9030649964d9353270 caused this, as further testing with ea7929f68635549e9546f5080e6050b9552b0d6a was able to finish and generated an HTML report file.

### Source code / Error logs
```
(node:5016) UnhandledPromiseRejectionWarning: MongoNetworkError: connection 0 to localhost:27017 timed out
    at Socket.<anonymous> (/home/kaching/zenbot/node_modules/mongodb/lib/core/connection/connection.js:339:7)
    at Object.onceWrapper (events.js:286:20)
    at Socket.emit (events.js:198:13)
    at Socket._onTimeout (net.js:443:8)
    at ontimeout (timers.js:436:11)
    at tryOnTimeout (timers.js:300:5)
    at listOnTimeout (timers.js:263:5)
    at Timer.processTimers (timers.js:223:10)
(node:5016) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:5016) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

```
"mongod -v":
```
2020-01-06T17:43:28.468+0000 D NETWORK  [main] fd limit hard:4096 soft:1024 max conn: 819
2020-01-06T17:43:28.476+0000 I CONTROL  [initandlisten] MongoDB starting : pid=5616 port=27017 dbpath=/data/db 64-bit host=irgendwas
2020-01-06T17:43:28.476+0000 I CONTROL  [initandlisten] db version v3.6.3
2020-01-06T17:43:28.476+0000 I CONTROL  [initandlisten] git version: 9586e557d54ef70f9ca4b43c26892cd55257e1a5
2020-01-06T17:43:28.476+0000 I CONTROL  [initandlisten] OpenSSL version: OpenSSL 1.1.1  11 Sep 2018
2020-01-06T17:43:28.476+0000 I CONTROL  [initandlisten] allocator: tcmalloc
2020-01-06T17:43:28.476+0000 I CONTROL  [initandlisten] modules: none
2020-01-06T17:43:28.476+0000 I CONTROL  [initandlisten] build environment:
2020-01-06T17:43:28.476+0000 I CONTROL  [initandlisten]     distarch: x86_64
2020-01-06T17:43:28.476+0000 I CONTROL  [initandlisten]     target_arch: x86_64
2020-01-06T17:43:28.476+0000 I CONTROL  [initandlisten] options: { systemLog: { verbosity: 1 } }
2020-01-06T17:43:28.477+0000 D -        [initandlisten] User Assertion: 29:Data directory /data/db not found. src/mongo/db/service_context_d.cpp 98
2020-01-06T17:43:28.500+0000 I STORAGE  [initandlisten] exception in initAndListen: NonExistentPath: Data directory /data/db not found., terminating
2020-01-06T17:43:28.500+0000 I CONTROL  [initandlisten] now exiting
2020-01-06T17:43:28.500+0000 I CONTROL  [initandlisten] shutting down with code:100
```

**Comment by DeviaVir** [comment]
`2020-01-06T17:43:28.500+0000 I STORAGE  [initandlisten] exception in initAndListen: NonExistentPath: Data directory /data/db not found., terminating`

**Comment by MarioHenze** [comment]
Okay sorry, the second log isn't helpful as it looks like the command uses some config defaults which are not satisfied with the package installed by apt.

After the installation of mongodb I just ran `sudo systemctl enable mongodb`. No errors were reported in `journalctl --unit=mongodb.service` and the database connection between zenbot and mongodb was working reliable before. Running the backfill commands on different days appended to the data respectively. Therefore I assume there is no inherent problem with my mongodb config.



cat /etc/mongodb.conf:
```
# mongodb.conf

# Where to store the data.
dbpath=/var/lib/mongodb

#where to log
logpath=/var/log/mongodb/mongodb.log

logappend=true

bind_ip = 127.0.0.1
#port = 27017

# Enable journaling, http://www.mongodb.org/display/DOCS/Journaling
journal=true

# Enables periodic logging of CPU utilization and I/O wait
#cpu = true

# Turn on/off security.  Off is currently the default
#noauth = true
#auth = true

# Verbose logging output.
#verbose = true

# Inspect all client data for validity on receipt (useful for
# developing drivers)
#objcheck = true

# Enable db quota management
#quota = true

# Set diagnostic logging level where n is
#   0=off (default)
#   1=W
#   2=R
#   3=both
#   7=W+some reads
#diaglog = 0

# Diagnostic/debugging option
#nocursors = true

# Ignore query hints
#nohints = true

# Disable the HTTP interface (Defaults to localhost:27018).
#nohttpinterface = true

# Turns off server-side scripting.  This will result in greatly limited
# functionality
#noscripting = true

# Turns off table scans.  Any query that would do a table scan fails.
#notablescan = true

# Disable data file preallocation.
#noprealloc = true

# Specify .ns file size for new databases.
# nssize = <size>

# Accout token for Mongo monitoring server.
#mms-token = <token>

# Server name for Mongo monitoring server.
#mms-name = <server-name>

# Ping interval for Mongo monitoring server.
#mms-interval = <seconds>

# Replication Options

# in replicated mongo databases, specify here whether this is a slave or master
#slave = true
#source = master.example.com
# Slave only: specify a single database to replicate
#only = master.example.com
# or
#master = true
#source = slave.example.com

# Address of a server to pair with.
#pairwith = <server:port>
# Address of arbiter server.
#arbiter = <server:port>
# Automatically resync if slave data is stale
#autoresync
# Custom size for replication operation log.
#oplogSize = <MB>
# Size limit for in-memory storage of op ids.
#opIdMem = <bytes>

# SSL options
# Enable SSL on normal ports
#sslOnNormalPorts = true
# SSL Key file and password
#sslPEMKeyFile = /etc/ssl/mongodb.pem
#sslPEMKeyPassword = pass
```

**Comment by jeremyloe** [comment]
Same issue... Just popped up recently. 

mongod -v
NonExistentPath: Data directory /data/db not found., terminating
sudo mkdir -p /data/db/
sudo chown `id -u` /data/db
mongod -v
Error about not enough space for journal files
Extended Primary Volume
mongod -v
It created journal files and ran tests. All good this time
Set connection.js timeouts back to original value. 360000

My first sim to go 30+ days since the issue started. Seems to be good now. Thanks

Update: It still dies if sim runs longer than 30 seconds. Suggestions?

my output of mongod -v
2020-03-01T23:43:35.145-0500 [DataFileSync] BackgroundJob starting: DataFileSync
2020-03-01T23:43:35.145-0500 shardObjTest passed
2020-03-01T23:43:35.145-0500 isInRangeTest passed
2020-03-01T23:43:35.145-0500 shardKeyTest passed
2020-03-01T23:43:35.145-0500 [initandlisten] MongoDB starting : pid=22616 port=27017 dbpath=/data/db 64-bit host=ubuntu
2020-03-01T23:43:35.145-0500 [initandlisten] db version v2.6.10
2020-03-01T23:43:35.145-0500 [initandlisten] git version: nogitversion
2020-03-01T23:43:35.145-0500 [initandlisten] OpenSSL version: OpenSSL 1.0.2g  1 Mar 2016
2020-03-01T23:43:35.145-0500 [initandlisten] build info: Linux lgw01-12 3.19.0-25-generic #26~14.04.1-Ubuntu SMP Fri Jul 24 21:16:20 UTC 2015 x86_64 BOOST_LIB_VERSION=1_58
2020-03-01T23:43:35.145-0500 [initandlisten] allocator: tcmalloc
2020-03-01T23:43:35.146-0500 [initandlisten] options: { systemLog: { verbosity: 1 } }
2020-03-01T23:43:35.155-0500 [initandlisten] flushing directory /data/db
2020-03-01T23:43:35.156-0500 [initandlisten] journal dir=/data/db/journal
2020-03-01T23:43:35.166-0500 [initandlisten] recover : no journal files present, no recovery needed
2020-03-01T23:43:35.166-0500 [initandlisten] flushing directory /data/db/journal
2020-03-01T23:43:35.171-0500 [initandlisten] flushing directory /data/db/journal
2020-03-01T23:43:35.190-0500 [initandlisten] opening db:  local
2020-03-01T23:43:35.255-0500 [initandlisten] enter repairDatabases (to check pdfile version #)
2020-03-01T23:43:35.255-0500 [initandlisten] 	local
2020-03-01T23:43:35.291-0500 [initandlisten] done repairDatabases
2020-03-01T23:43:35.291-0500 [initandlisten] opening db:  admin
2020-03-01T23:43:35.292-0500 [initandlisten] query admin.system.roles planSummary: EOF ntoreturn:0 ntoskip:0 nscanned:0 nscannedObjects:0 keyUpdates:0 numYields:0 locks(micros) W:99 r:48 nreturned:0 reslen:20 0ms
2020-03-01T23:43:35.292-0500 [ClientCursorMonitor] BackgroundJob starting: ClientCursorMonitor
2020-03-01T23:43:35.292-0500 [PeriodicTaskRunner] BackgroundJob starting: PeriodicTaskRunner
2020-03-01T23:43:35.292-0500 [TTLMonitor] BackgroundJob starting: TTLMonitor
2020-03-01T23:43:35.292-0500 [initandlisten] fd limit hard:1048576 soft:1024 max conn: 819
2020-03-01T23:43:35.292-0500 [IndexRebuilder] BackgroundJob starting: IndexRebuilder
2020-03-01T23:43:35.292-0500 [initandlisten] ERROR: listen(): bind() failed errno:98 Address already in use for socket: 0.0.0.0:27017
2020-03-01T23:43:35.292-0500 [initandlisten] ERROR:   addr already in use
2020-03-01T23:43:35.292-0500 [IndexRebuilder] opening db:  local
2020-03-01T23:43:35.292-0500 [initandlisten] create collection local.startup_log { size: 10485760, capped: true }
2020-03-01T23:43:35.302-0500 [initandlisten] command local.$cmd command: create { create: "startup_log", size: 10485760, capped: true } ntoreturn:1 keyUpdates:0 numYields:0  reslen:75 9ms
2020-03-01T23:43:35.302-0500 [initandlisten] insert local.startup_log ninserted:1 keyUpdates:0 numYields:0  0ms
2020-03-01T23:43:35.302-0500 [initandlisten] now exiting
2020-03-01T23:43:35.302-0500 [initandlisten] dbexit: 
2020-03-01T23:43:35.302-0500 [initandlisten] shutdown: going to close listening sockets...
2020-03-01T23:43:35.302-0500 [initandlisten] shutdown: going to flush diaglog...
2020-03-01T23:43:35.302-0500 [initandlisten] shutdown: going to close sockets...
2020-03-01T23:43:35.302-0500 [initandlisten] shutdown: waiting for fs preallocator...
2020-03-01T23:43:35.302-0500 [initandlisten] shutdown: lock for final commit...
2020-03-01T23:43:35.302-0500 [initandlisten] shutdown: final commit...
2020-03-01T23:43:35.304-0500 [initandlisten] shutdown: closing all files...
2020-03-01T23:43:35.304-0500 [initandlisten] closeAllFiles() finished
2020-03-01T23:43:35.304-0500 [initandlisten] journalCleanup...
2020-03-01T23:43:35.304-0500 [initandlisten] removeJournalFiles
2020-03-01T23:43:35.305-0500 [initandlisten] flushing directory /data/db/journal
2020-03-01T23:43:35.305-0500 [initandlisten] removeJournalFiles end
2020-03-01T23:43:35.305-0500 [initandlisten] shutdown: removing fs lock...
2020-03-01T23:43:35.306-0500 [initandlisten] dbexit: really exiting now

**Comment by Orwi53** [comment]
I get the same error during simulations: "`MongoNetworkError: connection 0 to localhost:27017 timed out`", etc.
The error abruptly stops the simulation after some time (in my case about 40 seconds). Some times, instead, the simulation stops with no additional output. 
Branch is "[unstable] Exchanges: update-products 5539", I am using NVM for changing node versions.

I did some tests using different node versions (in each case deleting "node_modules" first, then "npm install" and then `./zenbot.sh sim --days 20`, with poloniex.ETH-BTC in conf.js)

This happens with both node 10 and 11, however, **it doesn't happen with node 12**

I did those tests on a Ubuntu Mate 18.04 x86_64 laptop, and on a Raspberry PI 4 with Ubuntu 18.04 64-bit (aarch64) and 4 GB or ram. Same outcome in both machines (simulation stopping with node 10 and 11, but working with node 12)
